### PR TITLE
Implement SWAR indexOf byte search

### LINKED ISSUE #10731: Implement SWAR indexOf byte search
Currently Netty uses an expensive (GC-wise too) `ByteBufUtil::indexOf` implementation based on `ByteProcessor.IndexOfProcessor`  to save bound-checks to happen: this could be changed into what @richardstartin has implemented on https://richardstartin.github.io/posts/finding-bytes.

Same could be applied to `ByteBuf::indexOf` that's *not* using `ByteBufUtil::indexOf`(!!!)
@normanmaurer  Any idea why?

### PR DESCRIPTION
**Pull request by franz1981**
Motivation:

Faster indexOf

Modification:

Create generic SWAR indexOf that any ByteBuf implementation can use

Result:

Fixes #10731

### CODE REVIEW

**Comment by hyperxpro** [comment]
Any benchmark results for this?

**Comment by franz1981** [comment]
@hyperxpro 
> Any benchmark results for this?

That's a draft PR :)
Still working on the logic so doesn't make sense to bench it yet

**Comment by hyperxpro** [comment]
Makes sense. Tests failed too.

**Comment by hyperxpro** [comment]
```
13:46:56 [ERROR] /code/buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:442: Line is longer than 120 characters (found 128). [LineLength]
13:46:56 [ERROR] /code/buffer/src/test/java/io/netty/buffer/AbstractByteBufTest.java:2155: two or more consecutive empty lines [RegexpMultiline]
```

**Comment by franz1981** [comment]
@hyperxpro already fixed :+1:

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 1]
```diff
@@ -389,6 +389,89 @@ private static long compareUintBigEndianB(
...
+            SEARCH_PATTERNS = new long[256];
+            for (int i = 0; i < SEARCH_PATTERNS.length; ++i) {
+                final byte b = (byte) (i - 128);
+                SEARCH_PATTERNS[i] = compilePattern(b);
+            }
+        }
+
+        private static long compilePattern(byte byteToFind) {
+            long pattern = byteToFind & 0xFFL;
+            return pattern
```
Comment (richardstartin) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:406]: you can do this as `long pattern = (byteToFind & 0xFFL) * 0x1010101010101010L`

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 2]
```diff
@@ -389,6 +389,89 @@ private static long compareUintBigEndianB(
...
+                   | (pattern << 8)
```
Comment (richardstartin) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:407]: you can do this as long pattern = (byteToFind & 0xFFL) * 0x101010101010101L

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 3]
```diff
@@ -389,6 +389,89 @@ private static long compareUintBigEndianB(
...
+            tmp = ~(tmp | input | 0x7F7F7F7F7F7F7F7FL);
+            final int binaryPosition = leading? Long.numberOfLeadingZeros(tmp) : Long.numberOfTrailingZeros(tmp);
+            return binaryPosition >>> 3;
+        }
+
+        // TODO: save using any pattern for the 0 search case, if we care
+        private static int firstZeroByte(long word, long pattern, boolean leading) {
+            long tmp = (word & 0x7F7F7F7F7F7F7F7FL) + 0x7F7F7F7F7F7F7F7FL;
+            tmp = ~(tmp | word | 0x7F7F7F7F7F7F7F7FL);
+            final int binaryPosition = leading? Long.numberOfLeadingZeros(tmp) : Long.numberOfTrailingZeros(tmp);
```
Comment (richardstartin) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:432]: Since you're doing a little endian check elsewhere, I would consider making little and big endian variants of this method, just because it makes it harder to mix and match

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 4]
```diff
@@ -389,6 +389,81 @@ private static long compareUintBigEndianB(
...
 
+    private static final class SWARByteSearch {
+
+        private static final long[] SEARCH_PATTERNS;
+
+        static {
+            SEARCH_PATTERNS = new long[256];
+            for (int i = 0; i < SEARCH_PATTERNS.length; ++i) {
+                final byte b = (byte) (i - 128);
+                SEARCH_PATTERNS[i] = compilePattern(b);
```
Comment (richardstartin) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:400]: This is 2KB, but if you change this to the multiplication mentioned earlier, I don't see benefit from caching

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java (hunk 3 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:424]: Currently this one isn't used yet, but `firstAnyPattern` instead, but I got the good idea, I will do it for both, :+1:

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java (hunk 4 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:400]: That's better, it's now very cheap, great

**Comment by richardstartin** [comment]
I'm really interested to find out if this works well in netty. I think there are substring searches to look at too, which should benefit from the same wider access.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java (hunk 3 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:410]: Will do it right after breakfast :P

**Comment by franz1981** [comment]
> I'm really interested to find out if this works well in netty. I think there are substring searches to look at too, which should benefit from the same wider access.

Me too! And curious to check if the `TODO` I've added can make any difference too!

I'll come up with a bench soon, probably using yours on the article as a base: many thanks again for both the article and your help here bud :)

**Comment by franz1981** [comment]
@richardstartin @hyperxpro 

With a very unfair data set length and no optimization at all (including the right advice of @richardstartin re creating separate versions LE/BE of a method), I'm getting:
```
Benchmark                              (logPermutations)  (needleByte)  (seed)  (size)   Mode  Cnt    Score   Error   Units
ByteBufIndexOfBenchmark.bufferIndexOf                 16             0       1       8  thrpt   16   41.017 ± 0.418  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf                 16             0       1      16  thrpt   16   35.957 ± 0.556  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf                 16             0       1      32  thrpt   16   26.767 ± 0.450  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf                 16             0       1      64  thrpt   16   17.747 ± 0.403  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                   16             0       1       8  thrpt   16  100.340 ± 3.007  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                   16             0       1      16  thrpt   16   79.562 ± 2.257  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                   16             0       1      32  thrpt   16   60.568 ± 0.990  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                   16             0       1      64  thrpt   16   37.069 ± 0.776  ops/us
```
It seems promising :) @normanmaurer  wdyt?

**Comment by richardstartin** [comment]
@franz1981 it's probably worth testing with non random data too because it might often be quite predictable. I'm not really familiar with what this is used for in netty.

**Comment by hyperxpro** [comment]
@franz1981 Looks good!

**Comment by richardstartin** [comment]
@franz1981 so faster even when predictable? Nice!

**Comment by franz1981** [comment]
@richardstartin predicatable for the machine, not sure for the JIT eheh :P
But yes, it still seems a win-win, let me check what happen with (N*8) + M with M < 8 sized data too!

**Comment by franz1981** [comment]
And some results from mixed workloads with predictable data:
```
Benchmark                              (logPermutations)  (needleByte)  (seed)  (size)   Mode  Cnt    Score    Error   Units
ByteBufIndexOfBenchmark.bufferIndexOf                  4             0       1       7  thrpt   16  103.167 ±  3.625  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf                  4             0       1       8  thrpt   16  102.151 ±  3.264  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf                  4             0       1      11  thrpt   16   95.552 ±  0.905  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf                  4             0       1      15  thrpt   16   77.921 ±  0.147  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                    4             0       1       7  thrpt   16  105.330 ±  0.664  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                    4             0       1       8  thrpt   16   97.496 ± 12.752  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                    4             0       1      11  thrpt   16   93.064 ±  0.797  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                    4             0       1      15  thrpt   16   72.519 ±  1.747  ops/us
```

**Comment by franz1981** [comment]
While with progressive unpredictability:
```
Benchmark                              (logPermutations)  (needleByte)  (seed)  (size)   Mode  Cnt   Score   Error   Units
ByteBufIndexOfBenchmark.bufferIndexOf                 11             0       1      10  thrpt   32  81.586 ± 2.863  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf                 11             0       1      15  thrpt   32  62.978 ± 2.456  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf                 15             0       1      10  thrpt   32  39.627 ± 0.507  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf                 15             0       1      15  thrpt   32  36.405 ± 0.322  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                   11             0       1      10  thrpt   32  87.960 ± 0.571  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                   11             0       1      15  thrpt   32  70.150 ± 1.678  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                   15             0       1      10  thrpt   32  84.940 ± 0.813  ops/us
ByteBufIndexOfBenchmark.swarIndexOf                   15             0       1      15  thrpt   32  67.781 ± 2.355  ops/us
```

**Comment by franz1981** [comment]
I see that 
```java
final boolean useLE = nativeOrder == ByteOrder.LITTLE_ENDIAN;
``` 
Is a const condition: I would like to check if the compiler is treating it correctly as constant

**Comment by franz1981** [comment]
@richardstartin 
I've found the reason why the SWAR version is competitive for small sizes and predictable cases too: loop unrolling!
using a counted loop sized with `& 7` seems to give the JIT the confidence to loop unroll it, getting good results. 
I haven't had time to try this with OpenJ9 (that doesn't play well with compiler directives) or Graal, but seems fun :)

**Comment by franz1981** [comment]
I've improved the benchmarks and I'm getting some bad results with `direct ByteBufs`, need some investigations:
```
Benchmark                                   (direct)  (logPermutations)  (needleByte)  (noUnsafe)  (seed)  (size)   Mode  Cnt   Score   Error   Units
ByteBufIndexOfBenchmark.bufferIndexOf          false                  4             0       false       1       7  thrpt   16  98.510 ± 2.296  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf          false                  4             0       false       1      16  thrpt   16  69.948 ± 0.659  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf           true                  4             0       false       1       7  thrpt   16  97.303 ± 0.844  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf           true                  4             0       false       1      16  thrpt   16  76.216 ± 0.506  ops/us
ByteBufIndexOfBenchmark.swarIndexOf            false                  4             0       false       1       7  thrpt   16  92.764 ± 1.508  ops/us
ByteBufIndexOfBenchmark.swarIndexOf            false                  4             0       false       1      16  thrpt   16  81.231 ± 1.548  ops/us
ByteBufIndexOfBenchmark.swarIndexOf             true                  4             0       false       1       7  thrpt   16  82.591 ± 1.040  ops/us
ByteBufIndexOfBenchmark.swarIndexOf             true                  4             0       false       1      16  thrpt   16  77.056 ± 0.579  ops/us
```

**Comment by franz1981** [comment]
I've got to get rid of the costy `instanceof` check (now monomorphic with guard) due to wrappings.
Obviously this can't leak in a final PR because `unwrap` can become megamorphic in the real world and I haven't yet considered its cost vs `instanceof` chain. 
If the final code is going to replace the simple linear loop on `AbstractByteBuf::indexOf` , probably I don't need any `instanceof` check at all...

But that highlight some weird behaviour:
```
Benchmark                                   (direct)  (logPermutations)  (needleByte)  (noUnsafe)  (seed)  (size)   Mode  Cnt    Score   Error   Units
ByteBufIndexOfBenchmark.bufferIndexOf          false                  4             0       false       1       7  thrpt   16   99.473 ± 0.822  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf          false                  4             0       false       1      16  thrpt   16   68.845 ± 1.109  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf          false                  4             0       false       1      23  thrpt   16   56.401 ± 0.886  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf          false                 11             0       false       1       7  thrpt   16   66.893 ± 0.414  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf          false                 11             0       false       1      16  thrpt   16   61.697 ± 1.063  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf          false                 11             0       false       1      23  thrpt   16   45.084 ± 0.467  ops/us

ByteBufIndexOfBenchmark.swarIndexOf            false                  4             0       false       1       7  thrpt   16  105.112 ± 0.640  ops/us
ByteBufIndexOfBenchmark.swarIndexOf            false                  4             0       false       1      16  thrpt   16   78.831 ± 1.032  ops/us
ByteBufIndexOfBenchmark.swarIndexOf            false                  4             0       false       1      23  thrpt   16   52.058 ± 0.156  ops/us
ByteBufIndexOfBenchmark.swarIndexOf            false                 11             0       false       1       7  thrpt   16   87.866 ± 6.715  ops/us
ByteBufIndexOfBenchmark.swarIndexOf            false                 11             0       false       1      16  thrpt   16   79.690 ± 1.001  ops/us
ByteBufIndexOfBenchmark.swarIndexOf            false                 11             0       false       1      23  thrpt   16   51.793 ± 0.331  ops/us

ByteBufIndexOfBenchmark.bufferIndexOf           true                  4             0       false       1       7  thrpt   16   97.812 ± 0.261  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf           true                  4             0       false       1      16  thrpt   16   76.494 ± 0.255  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf           true                  4             0       false       1      23  thrpt   16   61.326 ± 0.298  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf           true                 11             0       false       1       7  thrpt   16   74.700 ± 0.709  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf           true                 11             0       false       1      16  thrpt   16   59.176 ± 0.483  ops/us
ByteBufIndexOfBenchmark.bufferIndexOf           true                 11             0       false       1      23  thrpt   16   42.687 ± 1.089  ops/us

ByteBufIndexOfBenchmark.swarIndexOf             true                  4             0       false       1       7  thrpt   16  116.720 ± 0.473  ops/us
ByteBufIndexOfBenchmark.swarIndexOf             true                  4             0       false       1      16  thrpt   16   86.459 ± 0.313  ops/us
ByteBufIndexOfBenchmark.swarIndexOf             true                  4             0       false       1      23  thrpt   16   53.437 ± 0.165  ops/us
ByteBufIndexOfBenchmark.swarIndexOf             true                 11             0       false       1       7  thrpt   16   82.703 ± 1.920  ops/us
ByteBufIndexOfBenchmark.swarIndexOf             true                 11             0       false       1      16  thrpt   16   79.196 ± 0.924  ops/us
ByteBufIndexOfBenchmark.swarIndexOf             true                 11             0       false       1      23  thrpt   16   52.810 ± 0.350  ops/us
```

The weird part is now
```
Benchmark                                   (direct)  (logPermutations)  (needleByte)  (noUnsafe)  (seed)  (size)   Mode  Cnt    Score   Error   Units
ByteBufIndexOfBenchmark.bufferIndexOf           true                  4             0       false       1      23  thrpt   16   61.326 ± 0.298  ops/us
ByteBufIndexOfBenchmark.swarIndexOf             true                  4             0       false       1      23  thrpt   16   53.437 ± 0.165  ops/u
```
Where the SWAR case is getting an hit

**Comment by franz1981** [comment]
I've manually unrolled the single byte case and now the perf is a bit better :)

Going soon to move to the final version of this PR

**Comment by franz1981** [comment]
This is the performance with the very last version:
NEW:
```
Benchmark                             (direct)  (logPermutations)  (needleByte)  (noUnsafe)  (pooled)  (seed)  (size)   Mode  Cnt    Score   Error   Units
ByteBufIndexOfBenchmark.indexOf           true                  4             0       false     false       1       7  thrpt   16  120.986 ± 0.916  ops/us
ByteBufIndexOfBenchmark.indexOf           true                  4             0       false     false       1      16  thrpt   16   77.810 ± 0.603  ops/us
ByteBufIndexOfBenchmark.indexOf           true                  4             0       false     false       1      23  thrpt   16   62.605 ± 0.513  ops/us
ByteBufIndexOfBenchmark.indexOf           true                  4             0       false     false       1      32  thrpt   16   64.717 ± 0.333  ops/us
ByteBufIndexOfBenchmark.indexOf           true                 11             0       false     false       1       7  thrpt   16   82.222 ± 3.057  ops/us
ByteBufIndexOfBenchmark.indexOf           true                 11             0       false     false       1      16  thrpt   16   74.278 ± 2.926  ops/us
ByteBufIndexOfBenchmark.indexOf           true                 11             0       false     false       1      23  thrpt   16   60.679 ± 0.561  ops/us
ByteBufIndexOfBenchmark.indexOf           true                 11             0       false     false       1      32  thrpt   16   60.645 ± 0.745  ops/us
```
OLD:
```
Benchmark                             (direct)  (logPermutations)  (needleByte)  (noUnsafe)  (pooled)  (seed)  (size)   Mode  Cnt   Score   Error   Units
ByteBufIndexOfBenchmark.indexOf           true                  4             0       false     false       1       7  thrpt   16  97.411 ± 0.703  ops/us
ByteBufIndexOfBenchmark.indexOf           true                  4             0       false     false       1      16  thrpt   16  76.275 ± 0.402  ops/us
ByteBufIndexOfBenchmark.indexOf           true                  4             0       false     false       1      23  thrpt   16  61.636 ± 0.280  ops/us
ByteBufIndexOfBenchmark.indexOf           true                  4             0       false     false       1      32  thrpt   16  48.430 ± 0.170  ops/us
ByteBufIndexOfBenchmark.indexOf           true                 11             0       false     false       1       7  thrpt   16  76.215 ± 0.844  ops/us
ByteBufIndexOfBenchmark.indexOf           true                 11             0       false     false       1      16  thrpt   16  60.042 ± 1.890  ops/us
ByteBufIndexOfBenchmark.indexOf           true                 11             0       false     false       1      23  thrpt   16  43.772 ± 0.693  ops/us
ByteBufIndexOfBenchmark.indexOf           true                 11             0       false     false       1      32  thrpt   16  29.927 ± 0.205  ops/us
```

I've tried to improve the case with `size 7` by manually loop-unrolling the checks and indeed I'm getting back a nice speedup.

**Comment by franz1981** [comment]
@normanmaurer @richardstartin  I need to replace the usage of JDK 8 `SplittableRandom` :( 
I cannot' use JDK 8 features on bench IIRC or am I wrong?

Let me know if I need to better summarize the improvements and what I've done here...
I suppose that protocols that need to search for some limiter will benefit from this, any idea which one I could check?

**Comment by franz1981** [comment]
TLDR it's delivering a consistent speedup in both predictable and un predictable workloads, regardless the input size

**Comment by franz1981** [comment]
FYI, I've left a 
```java
// TODO: save using any pattern for the 0 search case, if we care
private static int firstZeroByte(long word, boolean leading) {
    ...
```
In the code, but it seems that it won't bring any significant perf improvement vs the generic version, so it will likely removed for a final version

**Comment by franz1981** [comment]
Probably @linasm is interested to this too :)

**Comment by franz1981** [comment]
Not that urgent @normanmaurer / @chrisvest but any feedback on this? Let me know if the comments makes uneasy to follow what I've done here

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 5]
```diff
@@ -389,16 +389,132 @@ private static long compareUintBigEndianB(
...
+        private static int firstAnyPattern(long word, long pattern, boolean leading) {
+            long input = word ^ pattern;
+            long tmp = (input & 0x7F7F7F7F7F7F7F7FL) + 0x7F7F7F7F7F7F7F7FL;
+            tmp = ~(tmp | input | 0x7F7F7F7F7F7F7F7FL);
+            final int binaryPosition = leading? Long.numberOfLeadingZeros(tmp) : Long.numberOfTrailingZeros(tmp);
+            return binaryPosition >>> 3;
+        }
+
+        // TODO: save using any pattern for the 0 search case, if we care
+        private static int firstZeroByte(long word, boolean leading) {
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:407]: This method isn't used.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 6]
```diff
@@ -389,16 +389,132 @@ private static long compareUintBigEndianB(
...
+    }
+
+    private static int unrolledFirstIndexOf(AbstractByteBuf buffer, int fromIndex, int byteCount, byte value) {
+        assert byteCount > 0 && byteCount < 8;
+        if (buffer._getByte(fromIndex) == value) {
+            return fromIndex;
+        }
+        if (byteCount == 1) {
+            return -1;
+        }
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:422]: Did you also try a small counted loop?

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 7]
```diff
@@ -389,16 +389,132 @@ private static long compareUintBigEndianB(
         return 0;
     }
 
+    private static final class SWARByteSearch {
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:392]: I wonder if this class would make sense as a top-level ByteSearch class, and adopt more methods.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 8]
```diff
@@ -389,16 +389,132 @@ private static long compareUintBigEndianB(
...
+        }
+        return -1;
+    }
+
+    /**
+     * This is using a SWAR (SIMD Within A Register) batch read technique to minimize bound-checks and improve memory
+     * usage while searching for {@code value}.
+     */
+    static int firstIndexOf(AbstractByteBuf buffer, int fromIndex, int toIndex, byte value) {
+        fromIndex = Math.max(fromIndex, 0);
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:464]: Why would `fromIndex` ever be negative?

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 9]
```diff
@@ -389,16 +389,132 @@ private static long compareUintBigEndianB(
...
+        final int length = toIndex - fromIndex;
+        buffer.checkIndex(fromIndex, length);
+        if (!PlatformDependent.isUnaligned()) {
+            return linearFirstIndexOf(buffer, fromIndex, toIndex, value);
+        }
+        assert PlatformDependent.isUnaligned();
+        int offset = fromIndex;
+        final int byteCount = length & 7;
+        if (byteCount > 0) {
+            final int index = unrolledFirstIndexOf(buffer, fromIndex, byteCount, value);
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:477]: I suppose the first `linearFirstIndexOf` call for platforms that don't support unaligned access is because we have no idea what the underlying alignment of the buffer is. But do we need this unrolled call on platforms that do support unaligned access?

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java [hunk 10]
```diff
@@ -389,16 +389,132 @@ private static long compareUintBigEndianB(
...
+        for (int i = 0; i < longCount; i++) {
+            // use the faster available getLong
+            final long word = useLE? buffer._getLongLE(offset) : buffer._getLong(offset);
+            int index = SWARByteSearch.firstAnyPattern(word, pattern, isNative);
+            if (index < Long.BYTES) {
+                return offset + index;
+            }
+            offset += Long.BYTES;
+        }
+        return -1;
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:500]: Don't we need to handle a tail? Or is that what the initial unrolled call is for?

**Review by chrisvest — COMMENTED** [review]
I had some comments and questions.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java (hunk 10 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:500]: Exactly!

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java (hunk 9 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:477]: Yep, it won't improve much for unpredictable inputs but seems to help a lot the other cases

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java (hunk 8 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:464]: Copied from the original implementation, suppose can be removed now

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java (hunk 7 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:392]: Yes yes and yes :)

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java (hunk 6 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:422]: Yess with byte indexes too

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java (hunk 5 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:407]: Yep can be removed given that it won't improve things a lot

**Review by chrisvest — APPROVED** [review]

**Review by normanmaurer — APPROVED** [review]

**Comment by normanmaurer** [comment]
Thanks a lot @franz1981
